modules/dag-l1/src/main/scala/org/tessellation/dag/l1/infrastructure/block/BlockService.scala

```diff
@@ -0,0 +1,144 @@
+          parentHashToTip = hashedBlock.signed.value.parent.map(br => br.hash -> allTips.get(br.hash)).toList.toMap
+          notTipParents = parentHashToTip.collect { case (hash, None) => hash }.toList
+          tipParents = parentHashToTip.collect { case (hash, Some(value)) => hash -> value }
+          tipParentsSize = tipParents.size
+          tipParentsToRemove = tipParents.filter {
+            case (_, block) =>
+              allTips.size >= tipsConfig.minimumTipsCount + tipParentsSize &&
+                (
+                  allTips.size > tipsConfig.maximumTipsCount + tipParentsSize ||
+                    block.usages.size >= tipsConfig.maximumTipUsages - 1
+                )
+          }
+          tipParentsToLeave = tipParents -- tipParentsToRemove.keySet
+          remainingTips = allTips -- tipParentsToRemove.keySet
+          minTipHeight = remainingTips.values.map(_.block.signed.value.height.coerce).minOption.getOrElse(0L)
+          tryAddAsATip = remainingTips.size < tipsConfig.maximumTipsCount && minTipHeight < hashedBlock.signed.value.height.coerce
+          _ <- notTipParents.traverse(
+            blocks(_)
+              .modify[UpdateResult] {
+                case Some(ap @ AcceptedBlock(_, _, usages, false)) =>
+                  (ap.copy(usages = usages + hashedBlock.proofsHash).some, Success)
+                case unexpected =>
+                  (unexpected, UnexpectedStateShift(s"Expected Accepted block that IS NOT a tip! Got $unexpected."))
+              }
+              .handle
+          )
+          _ <- tipParentsToRemove.keys.toList.traverse(
+            blocks(_)
+              .modify[UpdateResult] {
+                case Some(ap @ AcceptedBlock(_, _, usages, true)) =>
+                  (ap.copy(usages = usages + hashedBlock.proofsHash, isTip = false).some, Success)
+                case unexpected =>
+                  (unexpected, UnexpectedStateShift(s"Expected Accepted block that IS a tip! Got $unexpected."))
+              }
+              .handle
+          )
+          _ <- tipParentsToLeave.keys.toList.traverse(
+            blocks(_)
+              .modify[UpdateResult] {
+                case Some(ap @ AcceptedBlock(_, _, usages, true)) =>
+                  (ap.copy(usages = usages + hashedBlock.proofsHash).some, Success)
+                case unexpected =>
+                  (
+                    unexpected,
+                    UnexpectedStateShift(
+                      s"Expected Accepted block that IS a tip and will remain a tip! Got $unexpected"
+                    )
+                  )
+              }
+              .handle
+          )
+          _ <- if (tryAddAsATip)
+            blocks(hashedBlock.proofsHash).modify {
+              case Some(ap @ AcceptedBlock(_, _, usages, false)) if usages.size <= tipsConfig.maximumTipUsages - 1 =>
+                (ap.copy(isTip = true).some, ())
+              case other => (other, ())
+            } else Sync[F].unit
```

Comment:
in general (not related to this piece of code) - is it possible to split this really big function into smaller ones? especially the ones that returns nothing (like this one selected). it can be simplified as
```scala
_ <- removeParents(tipParentsToRemove, ...)
_ <- keepParents(tipParentsToLeave, ...)
_ <- tryAddAsATip(...)
```

Reply:
Yeah, you are right! I will do that and it should become more readable.